velox/type/Type.h

```diff
@@ -501,6 +537,99 @@ class TypeBase : public Type {
+ * in Velox. The parameter "precision" represents the number of digits the
+ * Decimal Type can support and scale represents the number of digits to the
+ * right of the decimal point.
+ */
+class DecimalType : public Type {
+ public:
+  DecimalType(const uint8_t precision = 18, const uint8_t scale = 0)
+      : Type(
+            precision <= 18 ? TypeKind::SHORT_DECIMAL : TypeKind::LONG_DECIMAL),
+        typmod_(TYPMOD(precision, scale)),
+        isBigDecimal_(precision > 18) {
+    std::stringstream ss;
+    ss << TypeTraits<TypeKind::SHORT_DECIMAL>::name;
+    ss << "(";
+    ss << precision;
+    ss << ",";
+    ss << scale;
+    ss << ")";
+    decimalTypeStr_ = ss.str();
+  }
+
+  inline bool isPrimitiveType() const override {
+    return isBigDecimal_ ? TypeTraits<TypeKind::LONG_DECIMAL>::isPrimitiveType
+                         : TypeTraits<TypeKind::SHORT_DECIMAL>::isPrimitiveType;
+  }
+
+  inline bool isFixedWidth() const override {
+    return isBigDecimal_ ? TypeTraits<TypeKind::LONG_DECIMAL>::isFixedWidth
+                         : TypeTraits<TypeKind::SHORT_DECIMAL>::isFixedWidth;
+  }
+
+  inline const char* kindName() const override {
+    return isBigDecimal_ ? TypeTraits<TypeKind::SHORT_DECIMAL>::name
+                         : TypeTraits<TypeKind::LONG_DECIMAL>::name;
+  }
+
+  inline uint32_t size() const override {
+    // DECIMAL TypeSignature can have 2 children which can be
+    // long literals or variables to represent the precision and scale.
+    return NUM_DECIMAL_PARAMETERS;
+  }
+
+  inline bool operator==(const Type& other) const override {
+    return (
+        (TypeKind::SHORT_DECIMAL == other.kind() ||
+         TypeKind::LONG_DECIMAL == other.kind()) &&
+        this->typmod_ == dynamic_cast<const DecimalType*>(&other)->typmod());
+  }
+
+  bool compareParameter(const int pos, std::string parameter) const {
+    if (pos == 0) {
+      std::string precision = std::to_string(typmod_ >> 8);
+      return precision.compare(parameter) == 0;
+    } else if (pos == 1) {
+      std::string scale = std::to_string(((int8_t)typmod_));
+      return scale.compare(parameter) == 0;
+    }
+    throw std::invalid_argument{"Invalid parameter position"};
+  };
+
+  inline int typmod() const {
+    return typmod_;
+  }
+
+  inline const std::shared_ptr<const Type>& childAt(uint32_t) const override {
+    throw std::invalid_argument{"scalar type has no children"};
+  }
+
+  std::string toString() const override {
+    return decimalTypeStr_;
+  }
+
+  size_t cppSizeInBytes() const override {
+    return isBigDecimal_
+        ? sizeof(TypeTraits<TypeKind::LONG_DECIMAL>::NativeType)
+        : sizeof(TypeTraits<TypeKind::SHORT_DECIMAL>::NativeType);
+  }
+
+  folly::dynamic serialize() const override {
+    folly::dynamic obj = folly::dynamic::object;
+    obj["name"] = "Type";
+    obj["type"] = decimalTypeStr_;
+    return obj;
+  }
+
+ private:
+  const int typmod_;
+  const bool isBigDecimal_;
```

Comment:
use `TypeKind`?

Reply:
`TypeKind` is part of the base `class Type`

Reply:
That's right. Since you already have the TypeKind an extra isBigDecimal_ boolean is redundant.